fastly/diff.go

```diff
@@ -0,0 +1,110 @@
+package fastly
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+// KeyFunc calculates a key from an element
+type KeyFunc func(interface{}) (interface{}, error)
+
+// SetDiff diffs two sets using a key to identify which elements have been added, changed, removed or not modified.
+//
+// This object compares sets using Terraform's schema.Set methods (e.g. Difference() and Intersection())
+// so that the same differences displayed to the user are honoured here.
+//
+// SetDiff however is able to tell if two elements from two distinct sets have the same key. This is useful to detect
+// that an element should be updated instead of recreated on the remote server.
+type SetDiff struct {
+	keyFunc KeyFunc
+}
+
+// NewSetDiff creates a new SetDiff with a provided KeyFunc
+func NewSetDiff(keyFunc KeyFunc) *SetDiff {
+	return &SetDiff{
+		keyFunc: keyFunc,
+	}
+}
+
+// DiffResult diffs to Set objects and returns a DiffResult object containing the diffs.
+//
+// The DiffResult object will contain the element from newSet on the Modified field.
+func (h *SetDiff) Diff(oldSet, newSet *schema.Set) (*DiffResult, error) {
```

Comment:
The comment doesn't match the function name and should have thrown a lint error. So we should either rename the function to `DiffResult` or update the comment to `Diff`.
```suggestion
// Diff diffs two Set objects and returns a DiffResult object containing the diffs.
//
// The DiffResult object will contain the elements from newSet on the Modified field.
func (h *SetDiff) Diff(oldSet, newSet *schema.Set) (*DiffResult, error) {
```

Reply:
I did some refactoring and the comments got outdated. I'll fix the comments